File: src/main/java/com/founderrd/xda/XDAContainer.java
/**
 * Title:	XDAContainer
 * Description:	实现XDAInterface接口，底层使用了XDADocument和XDADecorator
 * Author:	杨天航(tianhang.yang@gmail.com)
 * version: 	1.0
 * time: 	2009.11.11
 */

package com.founderrd.xda;

import java.io.*;
import java.security.NoSuchAlgorithmException;
import java.util.Vector;

public class XDAContainer implements XDAInterface {
    final XDADecorator dec;
    private final XDADocument doc;

    public XDAContainer() {
        this.doc = new XDADocument();
        this.dec = new XDADecorator();
    }

    // XDA文档操作
    public void create(String filePath, byte bitsParam)
            throws FileNotFoundException, FooE {
        File xdaFile = new File(filePath);
        doc.create(xdaFile, (byte) 0x01, (byte) 0x00, (byte) 0x00, bitsParam);
    }

    public void create(final String filePath) throws FileNotFoundException,
            FooE {
        create(filePath, (byte) 0x04);
    }

    public void open(String filePath) throws IOException, FooE {
        File xdaFile = new File(filePath);
        if (!xdaFile.exists() || xdaFile.isDirectory())
            throw new FooE(FooE.INVALID_FILE_PATH);

        doc.open(xdaFile);
    }

    public void close() throws IOException {
        doc.close();
    }

    public void save(boolean nameTableCompress, boolean itemListCompress)
            throws NoSuchAlgorithmException, FooE, IOException {
        byte entryCompressMark = 0x00;
        if (nameTableCompress)
            entryCompressMark |= NAME_TABLE_MASK;
        if (itemListCompress)
            entryCompressMark |= COMPRESSITEMLIST_MASK;

        doc.saveChanges(entryCompressMark);
    }

    public void save() throws NoSuchAlgorithmException, FooE,
            IOException {
        save(true, true);
    }

    public void saveAs(String newXDAPath, byte bitsParam,
                       boolean nameTableCompress, boolean itemListCompress) throws NoSuchAlgorithmException, FooE, IOException {
        byte entryCompress = 0;
        if (nameTableCompress)
            entryCompress &= 0x01;
        if (itemListCompress)
            entryCompress &= 0x02;

        File newFile = new File(newXDAPath);
        doc.saveAs(newFile, (byte) 0x01, (byte) 0x00, (byte) 0x00, bitsParam, entryCompress);
    }

    public void saveAs(String newXDAPath) throws NoSuchAlgorithmException, FooE, IOException {
        saveAs(newXDAPath, (byte) 0x04, true, true);
    }

    // 编辑操作
    public void addFile(String pathInXDA, String targetFilePath, byte[] ecs)
            throws FooE {
        XDAInputStream inputSteam = dec.decorate(targetFilePath, ecs);

        doc.insertItem(pathInXDA, inputSteam, ecs);
    }

    public void addFile(String pathInXDA, byte[] targetDate, byte[] ecs)
            throws FooE {
        XDAInputStream inputSteam = dec.decorate(targetDate, ecs);
        doc.insertItem(pathInXDA, inputSteam, ecs);
    }

    public void replaceFile(String pathInXDA, String targetFilePath, byte[] ecs)
            throws FooE {
        XDAInputStream inputSteam = dec.decorate(targetFilePath, ecs);
        doc.replaceItem(pathInXDA, inputSteam, ecs);
    }

    public void replaceFile(String pathInXDA, byte[] targetDate, byte[] ecs)
            throws FooE {
        XDAInputStream inputSteam = dec.decorate(targetDate, ecs);
        doc.replaceItem(pathInXDA, inputSteam, ecs);
    }

    public void removeFile(String pathInXDA) throws FooE {
        doc.deleteItem(pathInXDA);
    }

    public void addDir(String pathInXDA, String dirPath, byte[] ecs)
            throws FooE {
        File dir = new File(dirPath);
        if (!dir.exists() || !dir.isDirectory())
            return;

        String parentPath = pathInXDA.concat("\\");
        doAddDir(parentPath, dir, ecs);
    }

    public void removeDir(String pathInXDA) {
        // 检查pathInXDA
        char ch = pathInXDA.charAt(pathInXDA.length() - 1);
        String pathDir = pathInXDA;
        if ((ch != '\\') && (ch != '/')) {
            pathDir = pathDir.concat("\\");
        }

        Vector<String> allLogicExistedItem = doc.getAllLogicExistedItem();
        for (String oneItemPath : allLogicExistedItem) {
            if (oneItemPath.startsWith(pathDir))
                try {
                    doc.deleteItem(oneItemPath);
                } catch (FooE e) {
                }
        }
    }

    public void extractFile(String pathInXDA, String targetFilePath)
            throws IOException, FooE {
        OutputStream targetStream = new FileOutputStream(targetFilePath);
        doc.extractItemStream(pathInXDA, targetStream, dec);
    }

    @Override
    public byte[] extractFile(String pathInXDA) throws IOException,
            FooE {
        ByteArrayOutputStream targetStream = new ByteArrayOutputStream();
        doc.extractItemStream(pathInXDA, targetStream, dec);
        return targetStream.toByteArray();
    }

    @Override
    public void extractDir(String dir) throws IOException, FooE {
        Vector<String> allLogicExistedItem = doc.getAllLogicExistedItem();
        for (String pathInXDA : allLogicExistedItem) {
            String targetFilePath = dir.concat(pathInXDA);
            createParentDir(targetFilePath);
            extractFile(pathInXDA, targetFilePath);
        }
    }

    @Override
    public void extractDir(String pathInXDA, String dir) throws FooE {
        if (XDADocument.PACKPATH_PATTERN.matcher(pathInXDA).matches())
            throw new FooE(FooE.INVALID_PACK_PATH);

        char ch = pathInXDA.charAt(pathInXDA.length() - 1);
        if ((ch != '\\') && (ch != '/')) {
            pathInXDA = pathInXDA.concat("\\");
        }
        ch = dir.charAt(dir.length() - 1);
        if ((ch == '\\') || (ch == '/')) {
            dir = dir.substring(0, dir.length() - 2);
        }

        Vector<String> allLogicExistedItem = doc.getAllLogicExistedItem();
        for (String oneItemPath : allLogicExistedItem) {
            if (oneItemPath.startsWith(pathInXDA)) {
                try {
                    String extractFile = dir.concat(oneItemPath);
                    createFile(extractFile);
                    extractFile(oneItemPath, extractFile);
                } catch (IOException | FooE e) {
                }
            }
        }
    }

    @Override
    public boolean validateFile() {
        return false;
    }

    @Override
    public int getMajorVersion() throws FooE {
        return doc.getHeader().getMajorVersion();
    }

    @Override
    public int getMinorVersion() throws FooE {
        return doc.getHeader().getMinorVersion();
    }

    private boolean createParentDir(String path) {
        File file = new File(path);
        File parent = file.getParentFile();
        return parent.mkdirs();
    }

    private void doAddDir(String parentPath, File dir, byte[] ecs)
            throws FooE {
        File[] children = dir.listFiles();
        for (File child : children) {
            if (child.isDirectory()) {
                doAddDir(parentPath.concat(child.getName()).concat("\\"),
                        child, ecs);
            } else {
                addFile(parentPath.concat(child.getName()), child
                        .getAbsolutePath(), ecs);
            }
        }
    }

    private boolean createFile(String filePath) throws IOException {
        File file = new File(filePath);
        File folder = file.getParentFile();
        folder.mkdirs();
        return file.createNewFile();
    }
}

File: src/main/kotlin/com/founderrd/xda/XDAContainer.kt
/**
 * Title:	XDAContainer
 * Description:	实现XDAInterface接口，底层使用了XDADocument和XDADecorator
 * Author:	杨天航(tianhang.yang@gmail.com)
 * version: 	1.0
 * time: 	2009.11.11
 */
package com.founderrd.xda

import com.founderrd.xda.constant.COMPRESS_ITEMLIST_MASK
import com.founderrd.xda.constant.COMPRESS_NAMETABLE_MASK
import com.founderrd.xda.util.and
import com.founderrd.xda.util.or
import java.io.*
import java.security.NoSuchAlgorithmException

class XDAContainer : XDA {

    val dec: XDADecorator = XDADecorator()
    private val doc: XDADocument = XDADocument()

    // XDA文档操作
    override fun create(filePath: String, bitsParam: Byte) {
        val xdaFile = File(filePath)
        doc.create(xdaFile, 0x01.toByte(), 0x00.toByte(), 0x00.toByte(), bitsParam)
    }

    override fun create(filePath: String) {
        create(filePath, 0x04.toByte())
    }

    override fun open(filePath: String) {
        val xdaFile = File(filePath)
        if (!xdaFile.exists() || xdaFile.isDirectory) throw XDAException(XDAException.INVALID_FILE_PATH)
        doc.open(xdaFile)
    }

    override fun close() {
        doc.close()
    }

    override fun save(compressNameTable: Boolean, compressItemList: Boolean) {
        var entryCompressMark: Byte = 0x00
        if (compressNameTable) entryCompressMark = entryCompressMark or COMPRESS_NAMETABLE_MASK
        if (compressItemList) entryCompressMark = entryCompressMark or COMPRESS_ITEMLIST_MASK
        doc.saveChanges(entryCompressMark)
    }

    override fun save() {
        save(compressNameTable = true, compressItemList = true)
    }

    override fun saveAs(
        path: String,
        bitsParam: Byte,
        compressNameTable: Boolean,
        compressItemList: Boolean
    ) {
        var entryCompress: Byte = 0
        if (compressNameTable) entryCompress = entryCompress and 0x01
        if (compressItemList) entryCompress = entryCompress and 0x02
        val newFile = File(path)
        doc.saveAs(newFile, 0x01.toByte(), 0x00.toByte(), 0x00.toByte(), bitsParam, entryCompress)
    }

    @Throws(NoSuchAlgorithmException::class, XDAException::class, IOException::class)
    override fun saveAs(path: String) {
        saveAs(path, 0x04.toByte(), compressNameTable = true, compressItemList = true)
    }

    override fun appendFile(internalPath: String, sourceFilePath: String, ecs: ByteArray) {
        val inputSteam = dec.decorate(sourceFilePath, ecs)
        doc.insertItem(internalPath, inputSteam, ecs)
    }

    override fun appendFile(internalPath: String, data: ByteArray, ecs: ByteArray) {
        val inputSteam = dec.decorate(data, ecs)
        doc.insertItem(internalPath, inputSteam, ecs)
    }

    override fun replaceFile(internalPath: String, sourceFilePath: String, ecs: ByteArray) {
        val inputSteam = dec.decorate(sourceFilePath, ecs)
        doc.replaceItem(internalPath, inputSteam, ecs)
    }

    override fun replaceFile(internalPath: String, data: ByteArray, ecs: ByteArray) {
        val inputSteam = dec.decorate(data, ecs)
        doc.replaceItem(internalPath, inputSteam, ecs)
    }

    override fun removeFile(internalPath: String) {
        doc.deleteItem(internalPath)
    }

    override fun appendDirectory(attachPath: String, directory: String, ecs: ByteArray) {
        val dir = File(directory)
        if (!dir.exists() || !dir.isDirectory) return
        val parentPath = attachPath + "\\"
        doAppendDirectory(parentPath, dir, ecs)
    }

    override fun removeDirectory(internalPath: String) {
        // 检查pathInXDA
        var pathDir = internalPath
        if (!(pathDir.endsWith('\\') || pathDir.endsWith('/'))) {
            pathDir += "\\"
        }
        for (path in doc.allLogicExistedItem) {
            if (path!!.startsWith(pathDir)) try {
                doc.deleteItem(path)
            } catch (e: XDAException) {
                continue
            }
        }
    }

    override fun extractFileTo(internalPath: String, destination: String) {
        val targetStream: OutputStream = FileOutputStream(destination)
        doc.extractItemStream(internalPath, targetStream, dec)
    }

    override fun extract(internalPath: String): ByteArray {
        val targetStream: OutputStream = ByteArrayOutputStream()
        doc.extractItemStream(internalPath, targetStream, dec)
        return (targetStream as ByteArrayOutputStream).toByteArray()
    }

    override fun extractTo(destination: String) {
        val allLogicExistedItem = doc.allLogicExistedItem
        for (pathInXDA in allLogicExistedItem) {
            val targetFilePath = destination + pathInXDA
            File(targetFilePath).parentFile.mkdirs()
            extractFileTo(pathInXDA, targetFilePath)
        }
    }

    override fun extractTo(internalPath: String, destination: String) {
        var pathInXDA = internalPath
        var dir = destination
        if (XDADocument.PACKPATH_PATTERN.matcher(pathInXDA)
                .matches()
        ) throw XDAException(XDAException.INVALID_PACK_PATH)
        var ch = pathInXDA[pathInXDA.length - 1]
        if (ch != '\\' && ch != '/') {
            pathInXDA += "\\"
        }
        ch = dir[dir.length - 1]
        if (ch == '\\' || ch == '/') {
            dir = dir.substring(0, dir.length - 2)
        }

        for (oneItemPath in doc.allLogicExistedItem) {
            if (oneItemPath.startsWith(pathInXDA)) {
                try {
                    val extractFile = dir + oneItemPath
                    createFile(extractFile)
                    extractFileTo(oneItemPath, extractFile)
                } catch (e: IOException) {
                    continue
                } catch (e: XDAException) {
                    continue
                }
            }
        }
    }

    override val contents: List<String>
        get() = doc.allLogicExistedItem

    override fun validateFile(): Boolean {
        return false
    }

    override val majorVersion: Int = doc.header.getMajorVersion().toInt()

    override val minorVersion: Int = doc.header.getMinorVersion().toInt()

    private fun doAppendDirectory(parentPath: String, directory: File, ecs: ByteArray) {
        val children = directory.listFiles() ?: return
        for (child in children) {
            if (child.isDirectory) {
                doAppendDirectory(
                    parentPath + child.name + "\\",
                    child, ecs
                )
            } else {
                appendFile(
                    parentPath + child.name, child
                        .absolutePath, ecs
                )
            }
        }
    }

    private fun createFile(filePath: String): Boolean {
        val file = File(filePath)
        val folder = file.parentFile
        folder.mkdirs()
        return file.createNewFile()
    }
}